### PR TITLE
Support for CodeGen2

### LINKED ISSUE #202: CodeGen2 compatibility
New version of CodeGen was released: https://github.com/salesforce/CodeGen2

### PR DESCRIPTION
**Pull request by Hoekz**
---
## 1. General Description
I and @richjohnson-wwt worked on taking #209 over the finish line for supporting CodeGen-2 models. @michaelfeil 's work was extremely helpful. His changes are present on this branch, as well as:

1. Small fix to the [converter script](https://github.com/Hoekz/fauxpilot/blob/main/converter/huggingface_gptj_convert.py#L126) to avoid crashing
2. Configs for each model and GPU count
3. Updated `setup.sh` to allow choosing and converting of [CodeGen 2 models](https://github.com/Hoekz/fauxpilot/blob/main/setup.sh)
4. Update to [documentation](https://github.com/Hoekz/fauxpilot/blob/main/documentation/server.md#dependencies) about how to get Docker and the NVIDIA Container Toolkit
5. Note in the [README](https://github.com/Hoekz/fauxpilot/blob/main/README.md#copilot-configuration-for-fauxpilot) about how to use with GitHub Copilot

## 2. Changes proposed in this PR:
1. #202
6. #209

Resolves: #202
See Also: #209

## 3. How to evaluate:
1. Describe how to evaluate such that it may be reproduced by the reviewer (s).
    1. Run `setup.sh`
    1. Choose Codegen 2 (option 3)
    1. Choose desired model size
1. Self assessment:
 - [X] Successfully build locally `docker-compose build`:
 - [X] Successfully tested the full solution locally

## 4. More Information

1. Infilling does not appear to work. We were also experimenting with [TabbyML](https://github.com/TabbyML/tabby) with CodeGen 2 and saw similar results that seemed to indicate there's something wrong with the model, not with how it is being converted or called. It is also worth noting the client created by Venthe does not support infilling anyway, so a PR or other work is necessary to test this further.
2. Is it worth upgrading? Our experience indicated that while CodeGen 2 supports more languages, the general performance is not noticeably better than CodeGen 1.

### CODE REVIEW

**Comment by Aincvy** [comment]
I tried some content in this PR, successfully ran codegen2-3_7B with 1 gpu. My GPU is 3060 12G, here's my execution process:

1. Download model: https://huggingface.co/michaelfeil/codegen2-3_7B-gptj/resolve/main/pytorch_model.bin
2. Execute bash command:
```bash
cd converter/
conda activate textgen
MODEL="codegen2-3_7B"
NUM_GPUS=1
python3 huggingface_gptj_convert.py -in_file ~/ai/text-generation-webui/models/michaelfeil_codegen2-3_7B-gptj -saved_dir ../models/${MODEL}-${NUM_GPUS}gpu/fastertransformer/1 -infer_gpu_num ${NUM_GPUS}
cp models/codegen2-3_7B-1gpu/fastertransformer/config.pbtxt ../models/codegen2-3_7B-1gpu/fastertransformer/
```
3. Some explanations:
  - textgen: A Python environment that can run https://github.com/oobabooga/text-generation-webui/.
  - Modify config.pbtxt's `is_half` value to 1,  Otherwise, it will give me an OOM error.
4. Return to fauxpilot's main directory and execute the command `sudo ./launch.sh` to run it.

Other things:
- I didn't use Docker images for downloading and converting models.
- Sharing a single conda environment may not be ideal, but it can save time and disk space.

**Comment by fdegier** [comment]
Thank you for your contribution and sorry for the delay. Will pick this up this weekend.

**Comment by Pandapip1** [comment]
Any updates?

**Comment by RemcoSchrijver** [comment]
I have been running this branch for a couple of days right now and seems to work nicely. However @Aincvy what platform you're running on? For me the 3.7B didn't run on my RTX 4070 12GB due to memory constraints, might be because my DE had to much overhead? Do you know how much memory the model used?

1B runs fine for me as mentioned and uses a bit over 4.5 GB or to be precise 4298MiB (Also ran fine on RTX 3070 Mobile).

**Comment by Aincvy** [comment]
@RemcoSchrijver Did you change the argument, isHalf?

Debian is my host system.

**Comment by RemcoSchrijver** [comment]
> @RemcoSchrijver Did you change the argument, isHalf?
> 
> Debian is my host system.
> 
> 
> 
> 

Ahh I missed that you were running at half precision, works nicely now!

File: documentation/server.md [hunk 1]
```diff
@@ -1,4 +1,30 @@
 
+## Dependencies
+
+When running on a new machine, you will need Docker and the NVIDIA Container Toolkit:
```
Comment (fdegier) [inline documentation/server.md]:
```suggestion
When running on a new machine, you will need Docker and the NVIDIA Container Toolkit, below are instruction for Linux, other methods of fulfilling the dependencies are possible:
```

**Pull request merged**